packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -4450,10 +4450,32 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin, Res
   ///    state restoration.
   @optionalTypeArgs
   Future<T?> push<T extends Object?>(Route<T> route) {
+    _debugCheckUsePagesAPI(route);
     _pushEntry(_RouteEntry(route, initialState: _RouteLifecycle.push));
     return route.popped;
   }
 
+  void _debugCheckUsePagesAPI(Route<dynamic> route) {
+    assert((){
+      if (route.settings is Page) {
```

Comment:
We should not allow people to push a page based route, this should just throw a flutter error if this if case is true.

Reply:
@chunhtai we also need to update the error msg, right?

Reply:
that is right

Reply:
How about 'A page-based route should not be added using the imperative API.' which copy from your comment.